### DISCUSSION TITLE
Get low level socket errors. (Specifically bind failure when listening on a server)

### DISCUSSION THREAD

**Discussion by SoylentGraham** [FirstGearGames/FishNet#442] [Feature Request]
I get notifications when my server fails to start, and the console gets the guessed-error
`Server failed to start. This usually occurs when the specified port is unavailable, be it closed or already in use.`

But to aid work, I want to know why it failed, (and OS will happily give you out real reasons for not binding to a port, eg. blocked <1000 on mac)

I'm happy to add some code to the transports to fetch proper errors, but there's no API to access that error, only state.

**Comment by FirstGearGames** [comment]
We can take a PR but this is strongly a per-transport situation, and not something we can implement universally.

**Reply by SoylentGraham** [reply]
> this is strongly a per-transport situation

I should probe into every kind of transport to find out why I couldnt connect/listen/etc ? (via the transport id from the `On Connection State Changed` ?)

No generic "it failed because <transport's explanation>" API?

**Comment by FirstGearGames** [comment]
Sockets do not usually tell you why they fail.

**Reply by SoylentGraham** [reply]
There's a whole ton of error codes specifically for it
https://www.ibm.com/docs/en/zos/2.3.0?topic=codes-sockets-return-errnos (winsock codes are almost identical)

https://github.com/NewChromantics/SoyLib/blob/master/src/SoySocket.cpp#L366
They're often context'y errors, but still a lot of very specific ones

- `bind()` fails, and gives you an error, you can pretty much tell `EADDRINUSE` means the port is already bound
- if bind fails on mac and gives you `EACCES` then its because port is <1024 and not root 
- You get network disconnections, buffer overflows, safe connections, routing changes errors

I don't think there's any ambigious errors when using sockets really.

**Comment by FirstGearGames** [comment]
You have to keep in mind FishNet rarely has direct access to the used transport. EG: Tugboat uses LiteNetLib which returns connect/disconnect/data callbacks, not actually sure if it provides specific reasoning in the callbacks. But even if it did the likeliness of all transports doing this is pretty low.

While this could maybe be implemented for some (again, have not looked), there's no way it would work for all. If anything the transport specifically should be printing to log when able, not FishNet itself.

**Comment by FirstGearGames** [comment]
Closing this as I do not believe time will be found to add this in specifically for each transport. I do encourage you to make a PR if you'd like to see this in any certain transport.

**Reply by SoylentGraham** [reply]
Yeah, I'll implement this, an API fishnet side, and all the low level stuff tugboat side.

I need to make changes to tugboat to be sandbox-compatible for mac app store anyway

**Reply by SoylentGraham** [reply]
There's essentially a finite reasons sockets will error, so I'll do an enum (or maybe a string anyway) fishnet side, instead of non-descript failure.
